core/trino-main/src/main/java/io/trino/split/ConnectorAwareSplitSource.java

```diff
@@ -80,7 +80,7 @@ public ListenableFuture<SplitBatch> getNextBatch(int maxSize)
             if (noMoreSplits) {
                 finished = true;
                 tableExecuteSplitsInfo = Optional.of(source.getTableExecuteSplitsInfo());
-                source = null;
+                close();
```

Comment:
`close` is interface method and should not be called when we ourselves are not done yet

let's extract close body as `closeSource()` and call it here & from `close`

Reply:
Done